llm.el

```diff
@@ -128,6 +131,52 @@ REQUIRED is whether this is required or not."
   type
   required)
 
+(cl-defstruct llm-media
+  "Contains media that can be sent as part of an interaction.
+
+MIME-TYPE is a string containing the mime type of the media.
+
+DATA is a (binary) string containing the data."
+  mime-type data)
+
+(defun llm--image-to-media (image)
+  "Convert an IMAGE object to an `llm-media' object."
+  (make-llm-media
+   :mime-type (pcase (image-property image :type)
+		('svg "image/svg+xml")
+		('webp "image/webp")
+		('png "image/png")
+		('gif "image/gif")
+		('tiff "image/tiff")
+		('jpeg "image/jpeg")
+		('xpm "image/x-xpixmap")
+		('xbm "image/x-xbitmap"))
+   :data (if-let ((data (image-property image :data))) data
+	   (with-temp-buffer
+	     (insert-file-contents-literally (image-property image :file))
+	     (buffer-string)))))
+
+(cl-defstruct llm-multipart
+  "A multipart message that can contain both text and media.
+
+PARTS is a list of the parts of the interaction.  Each element
+should be either a string for text, or a `llm-media' object for
```

Comment:
What does a multipart string mean?  Is that actually supported?

Reply:
IIRC Google explicitly mentions multipart text as a way to e.g. send a block of code separately from a block of instructions, as a better alternative to putting the code in quotes. I did a quick experiment to check that openai also accepts multipart text with no media, and it does work. Ollama explicitly does not support it, and they would have to change the api in order to do so. However, it is always possible to just concatenate the blocks together as a fallback, and I already do this for ollama.

Reply:
That's fascinating, thanks for the informative reply.  Can you share some of that explanation in the docstring here?